backend/seafile/seafile.go

```diff
@@ -0,0 +1,1132 @@
+			object := &Object{
+				fs:            f,
+				id:            entry.ID,
+				remote:        filePath,
+				pathInLibrary: filePathInLibrary,
+				size:          entry.Size,
+				modTime:       time.Now(),
+				libraryID:     libraryID,
+			}
+			entries = append(entries, object)
+		}
+	}
+	return entries
+}
+
+func (f *Fs) mkDir(ctx context.Context, dir string) error {
+	library, fullPath := f.splitPath(dir)
+	libraryID, err := f.getLibraryID(ctx, library)
+	if err != nil {
+		return err
+	}
+	return f.mkMultiDir(ctx, libraryID, fullPath)
+}
+
+func (f *Fs) mkMultiDir(ctx context.Context, libraryID, dir string) error {
+	// rebuild the path one by one
+	folder := ""
+	for _, singleDir := range splitPath(dir) {
+		folder = path.Join(folder, singleDir)
+		err := f.mkSingleDir(ctx, libraryID, folder)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (f *Fs) mkSingleDir(ctx context.Context, libraryID, dir string) error {
+	dirDetails, err := f.getDirectoryDetails(ctx, libraryID, dir)
+	if err == nil && dirDetails != nil {
+		// Don't fail if the directory exists
+		fs.Debugf(nil, "%s: Directory '%s' already exists", f.Name(), dir)
+		return nil
+	}
+	if err == fs.ErrorDirNotFound {
+		fs.Debugf(nil, "%s: Create directory '%s'", f.Name(), dir)
+		err = f.createDir(ctx, libraryID, dir)
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+	return err
+}
+
+func (f *Fs) getDirectoryEntries(ctx context.Context, libraryID, folder string, recursive bool) ([]api.DirEntry, error) {
+	if f.useOldDirectoryAPI {
+		return f.getDirectoryEntriesAPIv2(ctx, libraryID, folder)
+	}
+	return f.getDirectoryEntriesAPIv21(ctx, libraryID, folder, recursive)
+}
+
+// splitPath creates a slice of paths
+func splitPath(tree string) (paths []string) {
+	tree, leaf := path.Split(path.Clean(tree))
+	for leaf != "" && leaf != "." {
+		paths = append([]string{leaf}, paths...)
+		tree, leaf = path.Split(path.Clean(tree))
+	}
+	return
+}
+
+func (f *Fs) getCachedLibraries(ctx context.Context) ([]api.Library, error) {
+	f.librariesMutex.Lock()
+	defer f.librariesMutex.Unlock()
+
+	libraries, err := f.libraries.Get(librariesCacheKey, func(key string) (value interface{}, ok bool, error error) {
+		// Load the libraries if not present in the cache
+		libraries, err := f.getLibraries(ctx)
+		if err != nil {
+			return nil, false, err
+		}
+		return libraries, true, nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	// Type assertion
+	return libraries.([]api.Library), nil
+}
+
+func (f *Fs) newObject(ctx context.Context, remote string, size int64, modTime time.Time) *Object {
+	library, folder := f.splitPath(remote)
+	libraryID, _ := f.getLibraryID(ctx, library) // If error it means the library does not exist (yet)
+	object := &Object{
+		fs:            f,
+		remote:        remote,
+		libraryID:     libraryID,
+		pathInLibrary: folder,
+		size:          size,
+		modTime:       modTime,
+	}
+	return object
+}
+
+// Check the interfaces are satisfied
```

Comment:
That is a nice set of optional interfaces!

No `DirMove` though? So not possible to rename directories?

Reply:
DirMove is a difficult one. The API works this way:

`move path1/dir1 to path2` will give you `path2/dir1` or `path2/dir1 (1)` if `dir1` was existing, but the API is also not telling you that it renamed it to `dir1 (1)`.

With this behaviour it's also impossible to simply change the name of the directory.

I can look into it again with a fresh mind though

Reply:
I think I found a way to make the DirMove working. I need a bit more testing but it looks good so far 👍